Geometry rev

ldd69 opened the pull request:
I have slightly modified the geometry2D class. I have reproduced the examples that were in the original geometry.py file. Some small differences, that I have highlighted in my comments, identified by '###' in the new version of geometry.py. I still would like to improve the comments in order to have a clear explanation of the rationale.

wilsonmr committed:
added input_folder to environment

wilsonmr committed:
cast additional training input to string, so that it can be represented in lockfile yaml format

ldd69 committed:
rewritten geometry functions

ldd69 committed:
revised version of geometry

ldd69 committed:
Short programme to test the new geometry routines. Reproduces the examples of the previous code.

wilsonmr commented on anvil/geometry.py:
```diff
@@ -13,48 +13,68 @@ class ShiftsMismatchError(Exception):
 
     def get_shift(self, shifts: tuple = (1, 1), dims: tuple = (0, 1)) -> torch.Tensor:
-        r"""Given length, which refers to size of a 2D state (length * length)
+        """Given length, which refers to size of a 2D state (length * length)
```
r was purposely here because we have backslashes in the docstring which we explicitly don't want to treat as escape characters https://docs.python.org/3/reference/lexical_analysis.html?highlight=literals#string-and-bytes-literals

ldd69 replied:
ok. good! I thought I had added the extra 'r' by cutting and pasting stuff.

wilsonmr reviewed (commented):
looks nice, I'm going to go through and change a few things.. specifically the documentation for get shift is out of date

ldd69 commented:
Ok. I’ve gone back to your nomenclature, i.e. using ’split’. 

Please check the ‘###’ comments. L.

> On 30 Mar 2020, at 11:23, wilsonmr <notifications@github.com> wrote:
> 
> 
> @wilsonmr commented on this pull request.
> 
> looks nice, I'm going to go through and change a few things.. specifically the documentation for get shift is out of date
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/wilsonmr/anvil/pull/29#pullrequestreview-383719105>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACQVU3UDMBCREQZFHQO56P3RKBXJJANCNFSM4LVL5D7Q>.
> 


The University of Edinburgh is a charitable body, registered in
Scotland, with registration number SC005336.

wilsonmr committed:
made test script into pytest unit test, updated example docstring to be correct and consistent. Changed name of splitlexi to be consistent with which indices were being cast to which representation in splitcart

wilsonmr committed:
added to readme how to run the tests

wilsonmr commented:
Ok so I changed the test script into some proper unit tests. I also updated the get shift documentation so it is up to date and should be correct now.

Finally I changed the `splitlexi` to `lexisplit`. I might be wrong here but I feel like

splitcart fills the 2-D cartesian sites with the split indices and calls iteself split to cartesian

when I read splitlexi it rather seems like it fills the split representation sites with lexicographic indices and so in that sense it would be consistent to be called lexicographic to split. But perhaps I'm confusing which indices you are calling lexicographic?

wilsonmr committed:
reformatting with black

ldd69 commented:
Hi Michael

I don’t agree with your comment below. 

I think splitlexi fills the lexicographic array with the split indices… no? 

l. 

> On 30 Mar 2020, at 13:13, wilsonmr <notifications@github.com> wrote:
> 
> 
> Ok so I changed the test script into some proper unit tests. I also updated the get shift documentation so it is up to date and should be correct now.
> 
> Finally I changed the splitlexi to lexisplit. I might be wrong here but I feel like
> 
> splitcart fills the 2-D cartesian sites with the split indices and calls iteself split to cartesian
> 
> when I read splitlexi it rather seems like it fills the split representation sites with lexicographic indices and so in that sense it would be consistent to be called lexicographic to split. But perhaps I'm confusing which indices you are calling lexicographic?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/wilsonmr/anvil/pull/29#issuecomment-605962678>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACQVU3X5BCV34JJNR2KWQLDRKCEF3ANCNFSM4LVL5D7Q>.
> 


The University of Edinburgh is a charitable body, registered in
Scotland, with registration number SC005336.

wilsonmr commented:
well like I said, I either don't understand what lexicographic is referring to or it seemed inconsistent. I think here I have constructed a 2-D phi state with each element being it's lexicographic index

```python
>>> phi = torch.arange(16).view(4, 4)
>>> phi
tensor([[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11],
        [12, 13, 14, 15]])
>>> phi_flat = phi.flatten()
tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15])
```

So when we flatten it we just get a flat tensor from 0 to 15

If I'm right with this definition then when I index `phi_flat` with `g.lexisplit` then we will get the split state [phiA, phiB].

```python
>>> g.lexisplit
tensor([ 0,  2,  5,  7,  8, 10, 13, 15,  1,  3,  4,  6,  9, 11, 12, 14])
>>> phi_flat[g.flat_checker] # phiA
tensor([ 0,  2,  5,  7,  8, 10, 13, 15])
>>> phi_flat[~g.flat_checker] # phiB
tensor([ 1,  3,  4,  6,  9, 11, 12, 14])
```

Finally (with pytorch we can only index with tensors of dtype long, so I have to do some messing around here - also ignore the warning) if I get my state of lexicographic indices and index first with lexisplit and then splitcart we should expect we will get the the 2-D state back:

```python
>>> phi_flat[g.lexisplit][torch.tensor(g.splitcart, dtype=torch.long)]
__main__:1: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
tensor([[ 0,  1,  2,  3],
        [ 4,  5,  6,  7],
        [ 8,  9, 10, 11],
        [12, 13, 14, 15]])
```

so the name ordering now looks consistent to me, with the caveat that maybe I've misinterpreted what you mean by lexicographical

ldd69 commented:
> On 30 Mar 2020, at 16:50, wilsonmr <notifications@github.com> wrote:
> 
> 
> well like I said, I either don't understand what lexicographic is referring to or it seemed inconsistent. I think here I have constructed a 2-D phi state with each element being it's lexicographic index
> 
> >>> phi = torch.arange(16).view(4, 4)
> >>> phi
> tensor([[ 0,  1,  2,  3],
>         [ 4,  5,  6,  7],
>         [ 8,  9, 10, 11],
>         [12, 13, 14, 15]])
> >>> phi_flat = phi.flatten()
> tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15])
> So when we flatten it we just get a flat tensor from 0 to 15
> 
> 
Agree.
> If I'm right with this definition then when I index phi_flat with g.lexisplit then we will get the split state [phiA, phiB].
> 
> >>> g.lexisplit
> tensor([ 0,  2,  5,  7,  8, 10, 13, 15,  1,  3,  4,  6,  9, 11, 12, 14])
This is also correct, but we need the inverse function. I was assuming that the state is already in the split state, and we want to index it
with g.splitlexi in order to get the lexicographic representation. 

I think we need to be clear on this, maybe I am misunderstanding what you have in mind. 

L.  
> >>> phi_flat[g.flat_checker] # phiA
> tensor([ 0,  2,  5,  7,  8, 10, 13, 15])
> >>> phi_flat[~g.flat_checker] # phiB
> tensor([ 1,  3,  4,  6,  9, 11, 12, 14])
> Finally (with pytorch we can only index with tensors of dtype long, so I have to do some messing around here - also ignore the warning) if I get my state of lexicographic indices and index first with lexisplit and then splitcart we should expect we will get the the 2-D state back:
> 
> >>> phi_flat[g.lexisplit][torch.tensor(g.splitcart, dtype=torch.long)]
> __main__:1: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
> tensor([[ 0,  1,  2,  3],
>         [ 4,  5,  6,  7],
>         [ 8,  9, 10, 11],
>         [12, 13, 14, 15]])
> so the name ordering now looks consistent to me, with the caveat that maybe I've misinterpreted what you mean by lexicographical
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/wilsonmr/anvil/pull/29#issuecomment-606082529>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACQVU3WZ4MOQ2YQSTNXS7ODRKC5U3ANCNFSM4LVL5D7Q>.
> 


The University of Edinburgh is a charitable body, registered in
Scotland, with registration number SC005336.

wilsonmr commented:
I think it's easier to see with a 4x4 rather than 2x2:

we start with the splitcart, which is the split state index in the cartesian state sites:

```python
>>> g = Geometry2D(4)
>>> g.splitcart
tensor([[ 0,  8,  1,  9],
        [10,  2, 11,  3],
        [ 4, 12,  5, 13],
        [14,  6, 15,  7]], dtype=torch.int32)
```

Now we roll the state, in this case roll with shift=1 and dim=0 which gives a new state

```python
>>> rolled = g.splitcart.roll(1, 0) # roll state down one
>>> rolled
tensor([[14,  6, 15,  7],
        [ 0,  8,  1,  9],
        [10,  2, 11,  3],
        [ 4, 12,  5, 13]], dtype=torch.int32)
```

which is still a cartesian state and each site is now filled with the split state index of the neighbour above. If we flatten this state we have a lexicographically ordered state, each site is still filled with the split state index of the top neighbour

```python
>>> flat = rolled.flatten()
>>> flat
tensor([14,  6, 15,  7,  0,  8,  1,  9, 10,  2, 11,  3,  4, 12,  5, 13],
       dtype=torch.int32)
```

what we want is something which can index the split state (which is what the transformations return) and so we index with lexisplit, which means we now have a split state where each site is the split state index of the top neighbour

```python
>>> flat[g.lexisplit]
tensor([14, 15,  8,  9, 10, 11, 12, 13,  6,  7,  0,  1,  2,  3,  4,  5],
       dtype=torch.int32)
```

Now we can use this in an action where we want some nearest neighbour interaction schematically like

```python
split_state * split_state[result of get_shift]
```

We never need the inverse, because for the action and all observables as long as the `get_shift` returns indices for a split state which index sites upon shifting the cartesian state then we can calculate everything. If we wanted to save these states to a text file, then yes we would need an inverse. But so far we never had a use for this.

ldd69 commented:
Ignore my comment from last night. I think we are doing the same thing, we are just misunderstanding the naming. 
My thinking, using your notation above, is that:

phi[lexisplit]

means that:

i=0
while i<len(phi):
   phi[lexisplit][[i]] = phi[lexisplit[[i]]]
   i += 1

so lexisplit returns the index of a split state vector, and therefore I would call it splitlexi (i.e. the output of the function is the first part of the name, like in splitcart). 
Happy to go for the other convention as long as we agree on what is going on.

wilsonmr committed:
make the splitcart all dtype=long so they can be used as indices

wilsonmr commented:
> so lexisplit returns the index of a split state vector, and therefore I would call it splitlexi (i.e. the output of the function is the first part of the name, like in splitcart).
> Happy to go for the other convention as long as we agree on what is going on.

splitcart doesn't return a split state it returns a cartesian state, this is exactly what I meant by them not being consistent before. I resolved it by reading the docstrings which both started with a sentence like:

"Split to Cartesian.."

which I then used to keep one of the functions and swap the other one to be consistent. I can see an argument for calling it `cartsplit` which would then say something

"Cartesian state filled with split indices..."

but then I would insist that we change lexisplit back to split lexi and the docstring to

"Split state filled with lexigraphical indices.."

I think as it stands this is perfectly fine and I like the tests it adds, we can build upon these and think about getting a CI up and running
